test/hotspot/jtreg/compiler/vectorization/TestRoundVectorFloatRandom.java

```diff
@@ -0,0 +1,214 @@
+
+import java.util.Random;
+import compiler.lib.ir_framework.DontCompile;
+import compiler.lib.ir_framework.IR;
+import compiler.lib.ir_framework.IRNode;
+import compiler.lib.ir_framework.Run;
+import compiler.lib.ir_framework.RunInfo;
+import compiler.lib.ir_framework.Test;
+import compiler.lib.ir_framework.TestFramework;
+import compiler.lib.ir_framework.Warmup;
+
+public class TestRoundVectorFloatRandom {
+  private static final int ITERS  = 11000;
+  private static final int ARRLEN = 997;
+  private static final float ADD_INIT = -7500.f;
+
+  private static final float[] input = new float[ARRLEN];
+  private static final int[] res = new int[ARRLEN];
+  private static final Random rand = new Random();
+
+  public static void main(String args[]) {
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=8");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=16");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=32");
+  }
+
+  @DontCompile
+  int golden_round(float f) {
+    return Math.round(f);
+  }
+
+  @Test
+  @IR(counts = {IRNode.ROUND_VF, "> 0"},
+      applyIfPlatform = {"x64", "true"},
+      applyIfCPUFeature = {"avx", "true"})
+  @IR(counts = {IRNode.ROUND_VF, "> 0"},
+      applyIfPlatform = {"aarch64", "true"})
+  void test_round(int[] a0, float[] a1) {
+    for (int i = 0; i < a0.length; i+=1) {
+      a0[i] = Math.round(a1[i]);
+    }
+  }
+
+  @Run(test = "test_round")
+  @Warmup(ITERS)
+  void test_rounds(RunInfo runInfo) {
+    // Initialize
+    for (int i = 0; i < ARRLEN; i++) {
+      float val = ADD_INIT+(float)i;
+      input[i] = val;
+    }
+
+    test_round(res, input);
+    // skip test/verify when warming up
+    if (runInfo.isWarmUp()) {
+      return;
+    }
+
+    int errn = 0;
+    // a single precise float point is composed of 3 parts: sign/exponent/signicand
+    // exponent part of a float value
+    final int exponentStart = 0;
+    final int exponentShift = 23;
+    final int exponentWidth = 8;
+    final int exponentBound = 1 << exponentWidth;
+    // significant part of a float value
+    final int signicandWidth = exponentShift;
+    final int signicandBound = 1 << signicandWidth;
+    final int signicandNum = 128;
+
+    // prepare for data of significand part
+    int signicandValues[] = new int[signicandNum];
+    int signicandIdx = 0;
+    for (; signicandIdx < signicandWidth; signicandIdx++) {
+      signicandValues[signicandIdx] = 1 << signicandIdx;
+    }
+    for (; signicandIdx < signicandNum; signicandIdx++) {
+      signicandValues[signicandIdx] = rand.nextInt(signicandBound);
+    }
+    signicandValues[rand.nextInt(signicandNum)] = 0;
+
+    // generate input arrays for testing, then run tests & verify results
+
```

Comment:
There's no need for randomness or arrays or special values in the 32-bit case. You can, and should, test the entire 32-bit range in a few lines of code by using floatBitsToInt. You can make the test much faster by copy-and-pasting the library code for Math.round(float) and letting the JIT compile it.

Reply:
> There's no need for randomness or arrays or special values in the 32-bit case. You can, and should, test the entire 32-bit range in a few lines of code by using floatBitsToInt.

In previous discussion, there are several reasons why it's implemented in this way:
1. test the whole range of 32 bits is slow, and even slow for a 64 ranges double.
2. if it's too slow, then it's not feasible to make it an automatic test.
these are expected by @eme64.

> You can make the test much faster by copy-and-pasting the library code for Math.round(float) and letting the JIT compile it.

Previously, I had [this question](https://github.com/openjdk/jdk/pull/17753#issuecomment-1992519401), but from the point view of correctness of the golden value.
I think you make another point to change from @DontCompile to copying library java code. Thanks!

Reply:
As I remember, that's not the case in my local environment, i.e. it will take longer time.

Reply:
If it takes longer than a few seconds, there's something wrong with your computer or your compiler. It's only 4G tests. I get

```
          6,255.47 msec task-clock:u                     #    1.057 CPUs utilized             
    19,681,407,235      cycles:u                         #    3.146 GHz                         (95.55%)
   110,397,691,854      instructions:u                   #    5.61  insn per cycle              (95.55%)

```

Reply:
Ah, maybe you're right, I was using qemu for arm test.
Let me try it later on a real arm machine later.

Reply:
An update, I ran a modified Float test on a AWS `c7gn.2xlarge` EC2 instance (with 8 vCPU, 16G mem), and it tooks more than 20 minutes to finish. So it seems too long for an automatic test?

Reply:
[foo.zip](https://github.com/openjdk/jdk/files/14919238/foo.zip)

Runs in 6s.

Reply:
[foo.zip](https://github.com/openjdk/jdk/files/14919424/foo.zip)

Reply:
Does that work for you?

Reply:
Thanks for the sample code.

I modify the current test a bit by using your code, i.e. change from 2 level nested loop to single while loop as below, let's call it `new test`
```
  @Test
  static boolean test(int testInt) {
        float testFloat = Float.intBitsToFloat(testInt);
        return Math.round(testFloat) != golden_round(testFloat);
  }

  @Run(test = "test")
  static void test_rounds(RunInfo runInfo) {
    for (int i = 0; i < 1000; i++) {        
        test(i);
    }
    if (runInfo.isWarmUp()) {
        return;
    }
    boolean runTest = true; // modify here to have try.
    if (!runTest) return;
    int testInt = 0;
    boolean fail = false;
    do {
        fail |= test(testInt);
    } while (++testInt != 0);
    if (fail) {
        throw new RuntimeException();
    }
  }
```

It still took more than 5 minutes to finish the test; if I assign `runTest = false`, it will take seconds. So most of time is spent on the while loop in `test_rounds` with `@Run` annotation in new test, I'm not sure how the annotation @Run works, but seems that's the reason why it's slower than a pure while loop (in your sample code). But we need the annotations in the test (check below).
There are still some gaps between this new test and current test:
* we still not yet verify IR Node (`IRNode.ROUND_VF`); to verify it, we need to put the (part of) test into a nested loop, and put this loop in a function (`test_round` in current test), and annotate this function with `@IR` to verify the IR node.

Or maybe there are other ways to implement this test and qualify below requirements? Currently I'm not sure.
1. run in a minute, as we want it to be an automatic test,
2. verify Math.round (intrinsic) result,
3. verify IR node (`IRNode.ROUND_VF) generation,
4. make sure all the verification is done after the warmup.

Reply:
Yes, I see. The `@Run` annotation runs it all with the IR framework, which you don't really need for an exhaustive test over the 32-bit range.

I think I may have been rather misled by the "Add exhaustive tests for Math.round intrinsics" title, which this PR doesn't do. I strongly suggest that you do add an exhaustive test for the 32-bit range, without using the `@Run` annotation, just the bare code.

Reply:
And my test only works on the scalar version, of course.

Reply:
Do you mean add an exhaustive test for the 32-bit range for the vector instrinsic or scalar one?
The only issue I have for an exhaustive test for the 32-bit range is that it takes too long time to be an automatic test, and we don't want to add an manual test as it seldomly runs.

Reply:
> The only issue I have for an exhaustive test for the 32-bit range is that it takes too long time to be an automatic test, and we don't want to add an manual test as it seldomly runs.

But it takes a long time because you use the @Run attribute, surely. If you ran that test just as a test, without the IR framework, it'd be fine.

Reply:
Yes, you mght be right.
But previously @eme64 asked to change from a plain test to a test using IR framework.

So I'm not sure what to do now.

Reply:
@theRealAph out of office, so don't have much time to think this through. But maybe we want both, a slower IR test which ensures we have the desired IR (with random input values), and also a non-IR test that is faster and checks the correct results more exhaustively?

Reply:
Totally. Then, at least for float, you've got it all.

Reply:
Thanks. I've added new tests which check full 32 bits range for Float.